File: grpc/grpc-client-annotation-processor/src/main/java/io/koraframework/grpc/client/annotation/processor/GrpcClientExtensionFactory.java
package io.koraframework.grpc.client.annotation.processor;

import io.koraframework.kora.app.annotation.processor.extension.ExtensionFactory;
import io.koraframework.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

import static io.koraframework.grpc.client.annotation.processor.GrpcClassNames.abstractStub;

public class GrpcClientExtensionFactory implements ExtensionFactory {
    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var stubTypeElement = processingEnvironment.getElementUtils().getTypeElement(abstractStub.canonicalName());
        if (stubTypeElement == null) {
            return Optional.empty();
        }
        var stubErasure = processingEnvironment.getTypeUtils().erasure(stubTypeElement.asType());
        return Optional.of(new GrpcClientExtension(processingEnvironment, stubErasure));
    }
}

File: grpc/grpc-client-symbol-processor/src/main/kotlin/io/koraframework/grpc/client/ksp/GrpcClientExtensionFactory.kt
package io.koraframework.grpc.client.ksp

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import io.koraframework.grpc.client.ksp.GrpcClassNames.abstractCoroutineStub
import io.koraframework.grpc.client.ksp.GrpcClassNames.abstractStub
import io.koraframework.kora.app.ksp.extension.ExtensionFactory
import io.koraframework.kora.app.ksp.extension.KoraExtension
import io.koraframework.ksp.common.KspCommonUtils.getClassDeclarationByName

class GrpcClientExtensionFactory : ExtensionFactory {
    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        val abstractStub = resolver.getClassDeclarationByName(abstractStub)
        val abstractCoroutineStub = resolver.getClassDeclarationByName(abstractCoroutineStub)
        if (abstractStub == null && abstractCoroutineStub == null) {
            return null
        }
        return GrpcClientExtension(resolver, kspLogger, codeGenerator, abstractStub, abstractCoroutineStub)
    }
}
